### DISCUSSION TITLE
Specify Order Type via REST API

### DISCUSSION THREAD

**Discussion by jfaigan** [openTCS/opentcs#64] [Q&A] [answered]
Hello,

I have a scenario where certain vehicles need to be restricted to certain activities / areas of the plant and I was thinking to enforce this by using the Order Type field on the Transport Order.  Using the Order Type I would be able to identify which vehicle would be appropriate and then my system would generate the Transport Order with the correct Order Type.

However, there does not seem to be a way to specify an Order Type via the REST API.  Is this correct?

I tried to set the TO Properties collection with a Type property but this did not adjust the TO field but simply created a new Property (which is what I was expecting unfortunately)
![image](https://user-images.githubusercontent.com/14929031/203556597-2682f861-931b-4e05-8d22-285f9cbc9ea8.png)

Is there a way to specify this information via the API?  Is this something that is planned for a future release?

Regards,
Joel

**Comment by martingr** [comment, marked as answer]
Hi Joel, you are correct. Setting the type of a transport order is currently not possible via the openTCS web API (it is only possible via the Operations Desk application). It would be desirable to extend the web API with a corresponding option (and probably other ones that might be missing for transport orders), but at the moment there are no plans to do so.

**Reply by jfaigan** [reply]
Ok, thank you for the confirmation.  That is too bad.

Joel

**Reply by swltr** [reply]
@jfaigan Yes, the web API should definitely be improved.

The following might be worth noting here:

* The openTCS development team usually adds features (like making it possible to specify an order type via the web API) as part of their work in (funded) projects. Outside of such projects, there is very limited time to implement additional features, unfortunately. Pull Requests that help improve openTCS are quite welcome, of course.
* There may be alternatives to using an order type, depending on the actual situation/requirements:
  * You can set an _intended vehicle_ for a transport order. (This implies that only a very specific vehicle would take the transport order, though.)
  * You can use different destination operations (for location types and transport orders) in the different areas of the plant. (This would require the vehicle driver assigned to a vehicle to differentiate these destination operations, though.)

**Reply by jfaigan** [reply]
Hi Stefan,

Yes, thank you for the other options.  I will take a look.

I started looking at the source code in Github to see what would be required.  :-)  The biggest problem is that I am not a Java dev and do not have a dev environment set up.  The changes do not look to be that complicated.  I am a former .Net dev so the leap to Java is not that huge.

If I find some time, I will take a look to see if I can get set up and make the changes and then provide them to you for evaluation.  

Thanks for the help!

Joel

**Reply by swltr** [reply]
@jfaigan With 0882d5e10f1d62fd7aadd040f7b6910edce8485c, you can now also set transport order types via the web API.

**Reply by jfaigan** [reply]
> @jfaigan With [0882d5e](https://github.com/openTCS/opentcs/commit/0882d5e10f1d62fd7aadd040f7b6910edce8485c), you can now also set transport order types via the web API.

@swltr Ok! That is very good news!  I am looking forward to trying that out when it is released.  Thanks!
